core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```
@@ -386,45 +388,56 @@ private long checkDeletedAndGCed(VersionValue versionValue) {
-            throw new IndexFailedEngineException(shardId, index.type(), index.id(), e);
+        } catch (Exception e) {
+            result = new IndexResult(checkIfDocumentFailureOrThrow(index, e), index.version());
+        }
+        return result;
+    }
+
+    /**
+     * Inspects exception thrown when executing index or delete operations
+     *
+     * @return failure if the failure is a document specific failure (e.g. analysis chain failure)
+     * @throws ElasticsearchException if the failure caused the engine to fail
+     * (e.g. out of disk, lucene tragic event)
+     *
+     * Note: pkg-private for testing
+     */
+    final Exception checkIfDocumentFailureOrThrow(final Operation operation, final Exception failure) {
+        boolean isDocumentFailure;
+        try {
+            // When indexing a document into Lucene, Lucene distinguishes between environment related errors
+            // (like out of disk space) and document specific errors (like analysis chain problems) by setting
+            // the IndexWriter.getTragicEvent() value for the former. maybeFailEngine checks for these kind of
+            // errors and returns true if that is the case. We use that to indicate a document level failure
+            // and set the error in operation.setFailure. In case of environment related errors, the failure
+            // is bubbled up
+            isDocumentFailure = !((failure instanceof IllegalStateException || failure instanceof IOException)
+                    && maybeFailEngine(operation.operationType().getLowercase(), failure));
+        } catch (Exception inner) {
+            // we failed checking whether the failure can fail the engine, treat it as a persistent engine failure
+            isDocumentFailure = false;
+            failure.addSuppressed(inner);
+        }
+        if (isDocumentFailure) {
+            return failure;
+        } else {
+            ElasticsearchException exception = new ElasticsearchException(failure);
```

sorry but this is an absolute no-go. We worked so hard to get rid of this in so many places. We should just rethrow the original exception instead of hiding it in and ElasticsearchException. Yet, I think we can use a littel hack to make it work in this case:
```Java


    @SuppressWarnings("unchecked")
    static <T extends Throwable> void rethrow(Throwable t) throws T {
        throw (T) t;
    }
```

should allow you to just retrhow `failure` without declaring it. I think in this case it's OK